### PR TITLE
Allow version on BPMN diagrams to be set

### LINKED ISSUE #2465: Execution Platform Can Be Set for BPMN Diagrams
Similar to how the execution platform can be set for forms the execution platform version can be set for Platform and Cloud BPMN diagrams. The execution platform itself cannot be set before we introduce generic modeling (starting without execution platform).

### General / BPMN

* [x] Simplify execution platform chooser 
    * [x] only allow version to be changed
    * [x] We always show major.minor
    * [x] We account for version not being present (an `<unset>` state exists in the version chooser, if not present)
    * [x] Infer platform from meta-data > ns > default to Camunda 7
    * [x] Consistently set version to `major.minor.patch` in the document with patch defaulting to `0`
    * [x] Treat patch versions as minors (i.e. 7.16.1 is shown as 7.16)
* [x] Architecture: Extract finding execution platform into a shared space and ensure platform can always be retrieved
* [x] Design: Update [UI to match concept](https://github.com/camunda/camunda-modeler-design/issues/57)

### Forms

* [x] Always require or execution platform to be set upfront, use simplified engine profile chooser

### DMN

* [x] Keep as is / continues to work (readonly mode)

Follow up: Verify with Cloud team that they set the respective namespaces when creating and editing forms and BPMN diagrams (https://github.com/bpmn-io/internal-docs/issues/400)

---

Child of https://github.com/camunda/camunda-modeler/issues/2464

### PR DESCRIPTION
**Pull request by nikku**
__Work items__

* [x] Add comprehensive test coverage (cf. https://github.com/camunda/camunda-modeler/pull/2573/commits/24789bca8265e363e4b69f8a4059b5318d2b5d3a)
* [x] `<unset>` execution platform version is indicated in version select, changing it serializes with proper version + platform
* [x] Fix CI failing with recent modeler changes (https://github.com/camunda/camunda-modeler/runs/4324253953) 
* [x] Update [UI to match concept](https://github.com/camunda/camunda-modeler-design/issues/57)
* [x] Always use the latest execution platform version for new diagrams (https://github.com/camunda/camunda-modeler/pull/2573/commits/3c51f34de9b68392e5d97854e2f64e91e611b0d3)

----

Closes #2465 

----

capture EIWB6Z_optimized

---

__Build artifacts <a name="build-artifacts"></a>__

* https://camunda-modeler-on-demand.s3.eu-central-1.amazonaws.com/2465-version-bpmn/camunda-modeler-2465-version-bpmn-linux-x64.tar.gz
* https://camunda-modeler-on-demand.s3.eu-central-1.amazonaws.com/2465-version-bpmn/camunda-modeler-2465-version-bpmn-mac.dmg
* https://camunda-modeler-on-demand.s3.eu-central-1.amazonaws.com/2465-version-bpmn/camunda-modeler-2465-version-bpmn-mac.zip
* https://camunda-modeler-on-demand.s3.eu-central-1.amazonaws.com/2465-version-bpmn/camunda-modeler-2465-version-bpmn-win-ia32.zip
* https://camunda-modeler-on-demand.s3.eu-central-1.amazonaws.com/2465-version-bpmn/camunda-modeler-2465-version-bpmn-win-x64.zip

### CODE REVIEW

**Comment by barmac** [comment]
Perhaps we could make use of https://github.com/bpmn-io/execution-platform in this PR?

**Comment by philippfromme** [comment]
@nikku I have refactored the Engine Profile into a separate class (cf. https://github.com/camunda/camunda-modeler/pull/2573/commits/edad0795eb76b7435e40e36e0f0b399c7a15b9ca).

I've also investigated the failing test case (cf. https://github.com/camunda/camunda-modeler/blob/develop/client/src/app/__tests__/IntegrationSpec.js#L64). It's failing because `BpmnEditor.state.importing` is now being set to `false` _after_ the tab was updated so the new XML isn't imported. So far I haven't found a solution. It seems like we were just lucky the test wasn't failing before.

File: client/src/app/tabs/form/FormEditor.js [hunk 1]
```diff
@@ -65,6 +62,29 @@ export class FormEditor extends CachedComponent {
...
+      set: (engineProfile) => {
+        const { form } = this.getCached();
+
+        const root = form._state.schema;
+
+        const modeling = form.get('modeling');
+
+        modeling.editFormField(root, engineProfile);
+
+        this.setCached({ engineProfile });
```
Comment (nikku) [inline client/src/app/tabs/form/FormEditor.js:82]: I believe the caching logic can be moved to the helper, too. It is duplicated across all implementations.

File: client/src/app/tabs/form/FormEditor.js [hunk 2]
```diff
@@ -253,22 +264,20 @@ export class FormEditor extends CachedComponent {
...
 
-    const { engineProfile: cachedEngineProfile } = this.getCached();
+      this.engineProfile.setCached(engineProfile);
+    } catch (err) {
 
-    if (!engineProfilesEqual(engineProfile, cachedEngineProfile) && isKnownEngineProfile(engineProfile)) {
-      this.setCached({
-        engineProfile
-      });
+      // TODO
```
Comment (nikku) [inline client/src/app/tabs/form/FormEditor.js:273]: TODO: Fill with life. When could this throw?

Comment (philippfromme) [reply client/src/app/tabs/form/FormEditor.js:273]: I don't really like this at the moment. We only want to update the cached engine profile if we've got a known profile. There isn't a real error case. We just ignore unknown engine profiles.

File: client/src/app/tabs/form/FormEditor.js (hunk 1 above)
Comment (philippfromme) [reply client/src/app/tabs/form/FormEditor.js:82]: Let me have another look.

Comment (philippfromme) [reply client/src/app/tabs/form/FormEditor.js]: Done (cf. https://github.com/camunda/camunda-modeler/pull/2573/commits/913be8c6d9799c1175fa87f900da5d2870935d47#diff-2202f2da3a0e27c5a60258c0242730fac685169548dfc027103cb72378a34609R34).

File: client/src/app/tabs/form/FormEditor.js (hunk 2 above)
Comment (nikku) [reply client/src/app/tabs/form/FormEditor.js:271]: Shall we unset the cached version to `null` in the error case?

**Comment by nikku** [comment]
Ready for review. Please note that we slightly simplified the help text (only talking about _diagram_). Anything else would have meant additional technical complexity that I did not want to add, given we plan to drop the description in the future anyway.

**Comment by nikku** [comment]
As a follow up I'll give it five minutes to see if we can re-use https://github.com/camunda/camunda-modeler/pull/2573#issuecomment-979795964.

**Comment by andreasgeier** [comment]
@nikku pls provide an artifact to test (all entities)

**Comment by nikku** [comment]
Linked the artifacts. Will be available soon.

**Comment by nikku** [comment]
@smbea @andreasgeier Please add your review.

File: client/src/app/tabs/EngineProfile.less [hunk 3]
```diff
@@ -1,34 +1,3 @@
-:local(.EngineProfileSelection) {
```
Comment (smbea) [inline client/src/app/tabs/EngineProfile.less:1]: This is great👌. No more css worries here

**Review by smbea — APPROVED** [review]
Seems to work as expected and looks nice and clean

**Comment by andreasgeier** [comment]
When creating a BPMN platform diagram, version 7.15 is set.
When creating a Form for platform, version 7.16 is set.
7.16 is latest.

When creating a BPMN cloud diagram, version 1.1 is set.
When creating a Form for cloud, version 1.2 is set.
1.3 is latest.

Why not always the latest available version? Now it feels a bit random. I think we discussed setting latest (in the sense of latest that the Modeler is aware of) as prefered solution.

**Comment by nikku** [comment]
We'll make sure to consistently set the latest version.

**Comment by andreasgeier** [comment]
> Please note that we slightly simplified the help text (only talking about diagram). Anything else would have meant additional technical complexity that I did not want to add, given we plan to drop the description in the future anyway.

We already need to distinguish between platform and cloud descriptions, is the distinction between entities significant differently? Calling a form as a "diagram" is just wrong, imho. 

If there is no way of doing this as requested, I would suggest to replace the word "diagram" with "file".

**Comment by andreasgeier** [comment]
✅  Visually, it looks as expected.

**Comment by nikku** [comment]
> If there is no way of doing this as requested, I would suggest to replace the word "diagram" with "file".

Replacing it with "file" as you suggested. :+1:

**Comment by nikku** [comment]
https://github.com/camunda/camunda-modeler/pull/2573/commits/3c51f34de9b68392e5d97854e2f64e91e611b0d3 ensures we always use the latest execution platform version for new diagrams.

**Comment by nikku** [comment]
https://github.com/camunda/camunda-modeler/pull/2573/commits/8ec14b897e8173317bdaf06ee8f387c841badf18 ensures we use the generic `file` name, rather than diagram, to better match all kinds of versioned files.

**Comment by nikku** [comment]
New artifact build [is in progress](https://github.com/camunda/camunda-modeler/runs/4406547767?check_suite_focus=true).

@smbea @philippfromme Ready for another round of review.

**Comment by nikku** [comment]
@andreasgeier New build artifacts available that incorporate your review hints.

**Comment by andreasgeier** [comment]
I got input from technical writing. In a text paragraph, the sentence "Select the Camunda platform version." is correct. But as a title, it should be as "Select Camunda platform version". If possible at this late point, please consider.

**Review by andreasgeier — APPROVED** [review]
Both topics, set latest and "file" instead of "diagram" are fixed. Approved.

**Review by philippfromme — APPROVED** [review]
Tried to break it but couldn't. Looks great! 👏🏻

**Pull request merged**